Add safe_cast

HowardHinnant opened the pull request:
* Ensure no overflow in casts between enums and integral types.
* Fixes: RIPD-1702

This PR adds `safe_cast` which is a more restricted form of `static_cast`.  It only works between two integral types, or between an enumeration and integral type.  The destination type is restricted to be a type that can hold all values of the source.  For example no signed type can be cast to any unsigned type because the destination can't hold negative values from the source.

When the source or destination is an enumeration type, then the analysis takes place on the enumeration's underlying type.

This cast helps prevent errors that might potentially occur due to different platforms using different implementation-defined choices of underlying type for the same enumeration by catching unsafe casts at compile time.

In places where an unsafe cast is unavoidable, `static_cast` is used instead of `safe_cast`, which now serves as a visible marker that this is a place in the code where information loss might be possible (and is often prevented by the surrounding code logic, for example with run-time checks).

ripplelabs-jenkins commented:
## Jenkins Build Summary

Built from [this commit](https://github.com/ripple/rippled/commit/89983605b1e1bf8d780a478ca4416331431de929)

Built at __20181227 - 16:49:49__

### Test Results

Build Type | Log | Result | Status
---------- | --- | ------ | ------
msvc.Debug | [console](http://ci.ops.ripple.com/job/rippled-dev/job/PR-2814/1//console) | no test results, t: n/a | FAIL :red_circle: 
msvc.Debug, <br/>NINJA_BUILD=true | [console](http://ci.ops.ripple.com/job/rippled-dev/job/PR-2814/1//console) | no test results, t: n/a | FAIL :red_circle: 
gcc.Debug <br/>-Dcoverage=ON, <br/>TARGET=coverage_report, <br/>SKIP_TESTS=true | [logfile](http://ci.ops.ripple.com/job/rippled-dev/job/PR-2814/1//artifact/gcc.Debug_Dcoverage_ON_TARGET_coverage_report_SKIP_TESTS_true.txt) | no test results, t: 61s | FAIL :red_circle: 
docs, <br/>TARGET=docs | [logfile](http://ci.ops.ripple.com/job/rippled-dev/job/PR-2814/1//artifact/docs_TARGET_docs.txt) | 1 cases, 0 failed, t: 0m1s | PASS :white_check_mark: 
gcc.Release <br/>-Dassert=ON, <br/>MANUAL_TESTS=true | [logfile](http://ci.ops.ripple.com/job/rippled-dev/job/PR-2814/1//artifact/gcc.Release_Dassert_ON_MANUAL_TESTS_true.txt) | no test results, t: 108s | FAIL :red_circle: 
msvc.Debug <br/>-Dunity=OFF | [console](http://ci.ops.ripple.com/job/rippled-dev/job/PR-2814/1//console) | no test results, t: n/a | FAIL :red_circle: 
msvc.Release | [console](http://ci.ops.ripple.com/job/rippled-dev/job/PR-2814/1//console) | no test results, t: n/a | FAIL :red_circle: 
rpm | [logfile](http://ci.ops.ripple.com/job/rippled-dev/job/PR-2814/1//artifact/rpm.txt) | no test results, t: n/a | FAIL :red_circle: 
clang.Debug | [logfile](http://ci.ops.ripple.com/job/rippled-dev/job/PR-2814/1//artifact/clang.Debug.txt) | 1144 cases, 0 failed, t: 3m14s | PASS :white_check_mark: 
gcc.Debug | [logfile](http://ci.ops.ripple.com/job/rippled-dev/job/PR-2814/1//artifact/gcc.Debug.txt) | no test results, t: 62s | FAIL :red_circle: 
gcc.Debug <br/>-Dunity=OFF | [logfile](http://ci.ops.ripple.com/job/rippled-dev/job/PR-2814/1//artifact/gcc.Debug_Dunity_OFF.txt) | no test results, t: 52s | FAIL :red_circle: 
clang.Debug <br/>-Dunity=OFF | [logfile](http://ci.ops.ripple.com/job/rippled-dev/job/PR-2814/1//artifact/clang.Debug_Dunity_OFF.txt) | 1144 cases, 0 failed, t: 10m4s | PASS :white_check_mark: 
gcc.Release <br/>-Dassert=ON | [logfile](http://ci.ops.ripple.com/job/rippled-dev/job/PR-2814/1//artifact/gcc.Release_Dassert_ON.txt) | no test results, t: 109s | FAIL :red_circle: 
clang.Release <br/>-Dassert=ON | [logfile](http://ci.ops.ripple.com/job/rippled-dev/job/PR-2814/1//artifact/clang.Release_Dassert_ON.txt) | 1144 cases, 0 failed, t: 5m42s | PASS :white_check_mark: 
gcc.Debug <br/>-Dstatic=OFF | [logfile](http://ci.ops.ripple.com/job/rippled-dev/job/PR-2814/1//artifact/gcc.Debug_Dstatic_OFF.txt) | no test results, t: 61s | FAIL :red_circle: 
gcc.Debug <br/>-Dstatic=OFF -DBUILD_SHARED_LIBS=ON | [logfile](http://ci.ops.ripple.com/job/rippled-dev/job/PR-2814/1//artifact/gcc.Debug_Dstatic_OFF_DBUILD_SHARED_LIBS_ON.txt) | no test results, t: 61s | FAIL :red_circle: 
gcc.Debug, <br/>NINJA_BUILD=true | [logfile](http://ci.ops.ripple.com/job/rippled-dev/job/PR-2814/1//artifact/gcc.Debug_NINJA_BUILD_true.txt) | no test results, t: 47s | FAIL :red_circle: 
clang.Debug <br/>-Dunity=OFF -Dsan=address, <br/>PARALLEL_TESTS=false, <br/>DEBUGGER=false | [logfile](http://ci.ops.ripple.com/job/rippled-dev/job/PR-2814/1//artifact/clang.Debug_Dunity_OFF_Dsan_address_PARALLEL_TESTS_false_DEBUGGER_false.txt) | 1144 cases, 0 failed, t: 10m7s | PASS :white_check_mark: 
clang.Debug <br/>-Dunity=OFF -Dsan=undefined, <br/>PARALLEL_TESTS=false | [logfile](http://ci.ops.ripple.com/job/rippled-dev/job/PR-2814/1//artifact/clang.Debug_Dunity_OFF_Dsan_undefined_PARALLEL_TESTS_false.txt) | 1144 cases, 0 failed, t: 11m57s | PASS :white_check_mark:

HowardHinnant commented:
I'm going to need an assist with the links to errors above, or with a copy/paste of the error messages generated for VS (and thanks!).

ximinez commented:
@howardhinnant I'm working on that now and expect to have it to you soon :tm:.

HowardHinnant commented on src/ripple/ledger/ApplyView.h:
```diff
@@ -77,11 +79,11 @@ constexpr
 }
 
-static_assert(~tapRETRY == static_cast<ApplyFlags>(~0x20),
+static_assert(~tapRETRY == safe_cast<ApplyFlags>(static_cast<unsigned>(~0x20)),
```
In clang the expression `~0x20` is `int`, which can't be cast to `ApplyFlags` which has an underlying type of `unsigned`.  I tried `~0x20u`, but that is still an `int` expression.

ximinez replied:
No problem. I'm good with leaving this as-is. I didn't try in any other compiler, but that doesn't surprise me.

HowardHinnant commented on src/ripple/net/impl/RegisterSSLCerts.cpp:
```diff
@@ -59,7 +60,7 @@ registerSSLCerts(
     {
         ec = boost::system::error_code(
-            static_cast<int>(::ERR_get_error()),
+            safe_cast<int>(::ERR_get_error()),
```
Ok, thanks.  I'll change it back to `static_cast`.

HowardHinnant commented on src/ripple/protocol/digest.h:
```diff
@@ -257,7 +257,7 @@ void
 {
     beast::sha512_hasher h;
-    auto const digest = static_cast<
+    auto const digest = safe_cast<
```
Nice catch.  This should never have been changed, but I didn't detect it because of the `#ifdef _MSC_VER`.  Reverted, thanks.

ximinez replied:
You're welcome

nbougalis replied:
This code has the following comment:
>Call from main to fix magic statics pre-VS2015

We no longer support pre-VS2015 and "magic statics" should _magically work_ these days, which raises the question, is this workaround needed any longer? We should test, and if it isn't, then we should just remove it outright.

HowardHinnant replied:
Good point.  There is another such comment in src/ripple/net/impl/RPCCall.cpp .

Sounds like a good issue to address in a separate P/R.

HowardHinnant commented on src/test/nodestore/Timing_test.cpp:
```diff
@@ -121,7 +122,7 @@ class Sequence
         rngcpy (&value[0], value.size(), gen_);
         return NodeObject::createObject (
-            static_cast<NodeObjectType>(d_type_(gen_)),
+            safe_cast<NodeObjectType>(d_type_(gen_)),
```
`NodeObjectType : unsigned` is a good solution, done, thanks.

HowardHinnant commented on src/ripple/conditions/impl/Fulfillment.cpp:
```diff
@@ -116,7 +117,7 @@ Fulfillment::deserialize(
     std::unique_ptr<Fulfillment> f;
 
-    switch (static_cast<Type>(p.tag))
+    switch (safe_cast<Type>(static_cast<std::uint8_t>(p.tag)))
```
I'm glad you highlighted this.  After taking another look, I think the best thing to do is to perform the `switch` in the "widest type" and let the `switch default` take care of the errors.  Trying something, but not 100% positive VS will swallow it...

ximinez replied:
It worked great!

ximinez commented on src/ripple/net/impl/RegisterSSLCerts.cpp:
```diff
@@ -60,7 +59,7 @@ registerSSLCerts(
     {
         ec = boost::system::error_code(
-            safe_cast<int>(::ERR_get_error()),
+            static_cast<int>(::ERR_get_error()),
```
Looking at this again, would it be worth adding a range-check `assert` like you did with STLedgerEntry?

HowardHinnant replied:
I'm hesitant to put in an assert since this is trying to create an `error_code` that presumably some other code wants to handle, instead of terminate.  I've translated the out of range result into an `ERANGE`.  What do you think about that?

nbougalis commented on src/ripple/server/impl/BaseWSPeer.h:
```diff
@@ -214,7 +215,8 @@ send(std::shared_ptr<WSMsg> w)
             "closing slow client";
-        cr_.code = static_cast<boost::beast::websocket::close_code>(4000);
+        cr_.code = safe_cast<boost::beast::websocket::close_code>
+                               (static_cast<unsigned short>(4000u));
```
I think this is overkill. In fact, I think we could probably swap this to be:

    cr_.code = boost::beast::websocket::close_code::policy_error;
    cr_.reason = "Policy error: client is too slow.";

HowardHinnant replied:
Is the change of value from 4000 (not a `close_code`) to 1008 (`close_code::policy_error`) intentional?  I'm not sure of the meaning of 4000, and was leery of changing behavior.

The type of `cr_.code` is `uint16_t`, and the type of `close_code::policy_error` is `enum close_code : std::uint16_t`, so this would be an ideal case for `safe_cast` (which lives to do the enum <-> integral conversions safely):

```
cr_.code = safe_cast<std::uint16_t>(boost::beast::websocket::close_code::policy_error);
```

or:

```
cr_.code = safe_cast<decltype(cr_.code)>(boost::beast::websocket::close_code::policy_error);
```

nbougalis replied:
In this situation we _are_  actually closing the connection, so a close code is appropriate.

Looking at the WebSocket spec, codes in the 4000 range are "_reserved for private use and thus can't be registered.  Such codes can be used by prior agreements between WebSocket applications.  The interpretation of these codes is undefined by this protocol._" ([RFC 6455 7.4.2](https://tools.ietf.org/html/rfc6455#section-7.4.2))

The "policy error" message seemed appropriate (the "policy" in question being "no slow clients") however, RFC 6455 says that 1008 "_indicates that an endpoint is terminating the connection because it has *received a message* that violates its policy._" (emphasis mine).

However, we are in the "write" codepath here; we haven't "received" something, so I'm unsure if this is an appropriate error code to use. The RFC does say that 1008 "_is a generic status code that can be returned when there is no other more suitable status code_" but I'm no longer sure that it makes sense.

I'm fine with leaving this bit as is.

vinniefalco replied:
Analysis of theoretical close codes is always fun but in this case we should be pragmatic. None of the clients connecting to rippled's WebSocket interfaces are going to know what "4000" means. But they will all know what 1007 (`policy_error`) means.

HowardHinnant commented on src/ripple/net/impl/RegisterSSLCerts.cpp:
```diff
@@ -58,9 +58,14 @@ registerSSLCerts(
-            boost::asio::error::get_ssl_category());
+        auto x = ::ERR_get_error();
+        int i;
+        if (x > std::numeric_limits<int>::max())
```
With this information, I think it is best to revert these changes back to `static_cast`.  My changes are intended to harden our code against future accidental changes, not change existing behavior.

ximinez replied:
Oh, and, hey, I didn't notice before, but this is in a `#if BOOST_OS_WINDOWS` block starting on line 39. This function is really short without it.

HowardHinnant commented on src/ripple/nodestore/NodeObject.h:
```diff
@@ -29,6 +29,7 @@ namespace ripple {
 
 /** The types of node objects. */
 enum NodeObjectType
+    : unsigned
```
I've confirmed that the existing underlying type of `NodeObjectType` is `unsigned` on both gcc-5.2 and gcc HEAD, so this change cements the existing binary format (if there is an impact at all).

HowardHinnant commented on src/ripple/protocol/impl/STLedgerEntry.cpp:
```diff
@@ -40,6 +41,9 @@ STLedgerEntry::STLedgerEntry (Keylet const& k)
 
     set (format->elements);
 
+    assert(0u <= type_ &&
```
Good catch, done, thanks.

HowardHinnant commented on src/ripple/rpc/impl/ServerHandlerImp.cpp:
```diff
@@ -836,12 +836,11 @@ ServerHandlerImp::processRequest (Port const& port,
 
-    rpc_time_.notify (static_cast <beast::insight::Event::value_type> (
+    rpc_time_.notify (
         std::chrono::duration_cast <std::chrono::milliseconds> (
```
Actually the cast isn't needed at all because `beast::insight::Event::notify` can take any `duration`.  It internally rounds up to the next `millisecond`.

I removed the `static_cast` and kept in the `duration_cast` just to preserve existing behavior.  This call rounds towards zero to the `millisecond` instead of letting `notify` round up.  If we prefer the round up, we can simply remove the `duration_cast`.

HowardHinnant commented:
Just let me know if these changes are not what people want.

HowardHinnant commented on src/ripple/basics/safe_cast.h:
```diff
@@ -0,0 +1,78 @@
+    static_assert(sizeof(Dest) >= sizeof(Src) + not_same,
+        "Destination is too small to hold all values of source");
+    return static_cast<Dest>(s);
+}
```
Thanks for the thorough testing!

ximinez replied:
Controversial opinion: I think we should include @scottschurr's tests. Reasons:
1. It verifies that everything is working as expected on all of our current supported platforms.
2. All the types listed have explicit sizes, so if a test did fail, there would be something unusual enough that we would _really_ want to know about it.
3. I'd be willing to bet that we have code all over the place that assumes these types have the size relationships that they do, even without explicit casting. If that were ever to not be true, similar to 2, it would be unusual enough that we'd want to know about it.

HowardHinnant replied:
The only way I know of to include these tests in our unit tests is to move the `static_assert`s to the `enable_if`s.  This will sacrifice the `static_assert` messages which will fire as we develop, even though `safe_cast` is developed and proven only once.

Though if you know of a way to apply these tests and not sacrifice the `static_assert` messages, that would be the best of both worlds.

HowardHinnant replied:
As an example of the `static_assert` messages adding usability:

Consider the scenario where we change the underlying type of `Type` to `std::int8_t` because we want to add a negative member:

https://github.com/ripple/rippled/blob/develop/src/ripple/conditions/Condition.h#L37

This will cause an error far away, here:

https://github.com/ripple/rippled/pull/2814/commits/7d0c01785fa595ad640fecd810529cd345ebd335#diff-9634052245df9b2eee790a9804088d25R123

If we add the tests, the error message will be something like:

`No matching call to safe_cast.`

As it is now, the error message is:

`Cannot cast signed to unsigned`

scottschurr replied:
I agree with Howard that the `static_assert`s significantly improve usability.

Something I've wished for but have never seen is a CI phase that compiles a bunch of files with the expectation that each compilation will fail.  Then all of the cases that we expect to fail to compile could be put in separate files and run during that CI phase.  I've never seen any  framework that supports that model.  So I shrug and do occasional manual testing.

HowardHinnant replied:
I wrote myself such a test framework for libcxx and date.  But it doesn't fit well into the rippled model where tests are part of the application binary.  What I wrote is a simplistic bash shell combined with a policy that tests which are expected to fail have filenames that end with `.fail.cpp`.  The bash script just runs over them and marks them as fail if they compile.  It doesn't even have the ability to check that they failed for the right reason.

HowardHinnant commented on src/ripple/ledger/ApplyView.h:
```diff
@@ -20,13 +20,15 @@
 namespace ripple {
 
 enum ApplyFlags
+    : unsigned
```
I think this is a good suggestion, done.

HowardHinnant commented on src/ripple/protocol/impl/tokens.cpp:
```diff
@@ -266,7 +267,7 @@ decodeBase58Token (std::string const& s,
 
     // The type must match.
-    if (type != static_cast<TokenType>(ret[0]))
+    if (type != safe_cast<TokenType>(static_cast<std::uint8_t>(ret[0])))
```
Done except I retained the implicit conversion syntax from the return type of `decodeBase58` because I don't want to accidentally invoke an explicit conversion from `decodeBase58` should its return type change in the future.

HowardHinnant commented:
I've addressed Scott's comments, retested and squashed everything.  Marking as Passed.  If anyone has any further concerns or comments, just let me know.  Thanks everyone for the good reviews.

HowardHinnant commented:
Rebased to 1.2.0-b10.
